File: step_15_graphs/15.2_bfs_dfs/P010_word_ladder_i/Solution.java
import java.util.*;

// ============================================================
// APPROACH 1: BRUTE FORCE - BFS, generate all neighbors naively
// Time: O(N * L^2)  |  Space: O(N * L)
// For each word in queue, try replacing each character with a-z
// ============================================================
class BruteForce {
    public static int solve(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (!wordSet.contains(endWord)) return 0;
        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);
        Set<String> visited = new HashSet<>();
        visited.add(beginWord);
        int steps = 1;
        while (!queue.isEmpty()) {
            int size = queue.size();
            steps++;
            for (int i = 0; i < size; i++) {
                String word = queue.poll();
                char[] arr = word.toCharArray();
                for (int j = 0; j < arr.length; j++) {
                    char orig = arr[j];
                    for (char c = 'a'; c <= 'z'; c++) {
                        if (c == orig) continue;
                        arr[j] = c;
                        String next = new String(arr);
                        if (next.equals(endWord)) return steps;
                        if (wordSet.contains(next) && !visited.contains(next)) {
                            visited.add(next);
                            queue.offer(next);
                        }
                    }
                    arr[j] = orig;
                }
            }
        }
        return 0;
    }
}

// ============================================================
// APPROACH 2: OPTIMAL - BFS with wordSet for O(1) lookup
// Time: O(N * L * 26)  |  Space: O(N * L)
// Remove visited words from set to avoid revisiting
// ============================================================
class Optimal {
    public static int solve(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (!wordSet.contains(endWord)) return 0;
        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);
        wordSet.remove(beginWord);
        int steps = 1;
        while (!queue.isEmpty()) {
            steps++;
            for (int size = queue.size(); size > 0; size--) {
                String word = queue.poll();
                char[] arr = word.toCharArray();
                for (int j = 0; j < arr.length; j++) {
                    char orig = arr[j];
                    for (char c = 'a'; c <= 'z'; c++) {
                        if (c == orig) continue;
                        arr[j] = c;
                        String next = new String(arr);
                        if (next.equals(endWord)) return steps;
                        if (wordSet.contains(next)) {
                            wordSet.remove(next);
                            queue.offer(next);
                        }
                    }
                    arr[j] = orig;
                }
            }
        }
        return 0;
    }
}

// ============================================================
// APPROACH 3: BEST - Bidirectional BFS
// Time: O(N * L * 26)  |  Space: O(N * L)
// Expand from both ends; much faster in practice (reduces search space)
// ============================================================
class Best {
    public static int solve(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (!wordSet.contains(endWord)) return 0;
        Set<String> beginSet = new HashSet<>(), endSet = new HashSet<>();
        beginSet.add(beginWord);
        endSet.add(endWord);
        wordSet.remove(endWord);
        int steps = 1;
        while (!beginSet.isEmpty()) {
            steps++;
            // Always expand the smaller set
            if (beginSet.size() > endSet.size()) {
                Set<String> tmp = beginSet; beginSet = endSet; endSet = tmp;
            }
            Set<String> nextSet = new HashSet<>();
            for (String word : beginSet) {
                char[] arr = word.toCharArray();
                for (int j = 0; j < arr.length; j++) {
                    char orig = arr[j];
                    for (char c = 'a'; c <= 'z'; c++) {
                        if (c == orig) continue;
                        arr[j] = c;
                        String next = new String(arr);
                        if (endSet.contains(next)) return steps;
                        if (wordSet.contains(next)) {
                            wordSet.remove(next);
                            nextSet.add(next);
                        }
                    }
                    arr[j] = orig;
                }
            }
            beginSet = nextSet;
        }
        return 0;
    }
}

public class Solution {
    public static void main(String[] args) {
        List<String> wordList1 = Arrays.asList("hot","dot","dog","lot","log","cog");
        System.out.println("Test 1 (hit->cog): expected 5");
        System.out.println("  BruteForce = " + BruteForce.solve("hit", "cog", new ArrayList<>(wordList1)));
        System.out.println("  Optimal    = " + Optimal.solve("hit", "cog", new ArrayList<>(wordList1)));
        System.out.println("  Best       = " + Best.solve("hit", "cog", new ArrayList<>(wordList1)));

        List<String> wordList2 = Arrays.asList("hot","dot","dog","lot","log");
        System.out.println("Test 2 (hit->cog, no cog in list): expected 0");
        System.out.println("  Optimal    = " + Optimal.solve("hit", "cog", new ArrayList<>(wordList2)));

        List<String> wordList3 = Arrays.asList("a","b","c");
        System.out.println("Test 3 (a->c): expected 2");
        System.out.println("  Optimal    = " + Optimal.solve("a", "c", new ArrayList<>(wordList3)));
    }
}

File: step_15_graphs/15.2_bfs_dfs/P010_word_ladder_i/solution.py
"""Problem: Word Ladder I
Difficulty: HARD | XP: 50

Find shortest transformation sequence from beginWord to endWord.
Each step changes exactly one letter, and intermediate words must be in wordList.
"""
from collections import deque
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE - BFS with visited set
# Time: O(N * L * 26)  |  Space: O(N * L)
# ============================================================
def brute_force(beginWord: str, endWord: str, wordList: List[str]) -> int:
    word_set = set(wordList)
    if endWord not in word_set:
        return 0
    queue = deque([(beginWord, 1)])
    visited = {beginWord}
    while queue:
        word, steps = queue.popleft()
        for i in range(len(word)):
            for c in 'abcdefghijklmnopqrstuvwxyz':
                if c == word[i]:
                    continue
                new_word = word[:i] + c + word[i+1:]
                if new_word == endWord:
                    return steps + 1
                if new_word in word_set and new_word not in visited:
                    visited.add(new_word)
                    queue.append((new_word, steps + 1))
    return 0


# ============================================================
# APPROACH 2: OPTIMAL - BFS, remove from set instead of visited set
# Time: O(N * L * 26)  |  Space: O(N * L)
# Removing from wordSet is cleaner and slightly faster
# ============================================================
def optimal(beginWord: str, endWord: str, wordList: List[str]) -> int:
    word_set = set(wordList)
    if endWord not in word_set:
        return 0
    queue = deque([beginWord])
    word_set.discard(beginWord)
    steps = 1
    while queue:
        steps += 1
        for _ in range(len(queue)):
            word = queue.popleft()
            for i in range(len(word)):
                for c in 'abcdefghijklmnopqrstuvwxyz':
                    new_word = word[:i] + c + word[i+1:]
                    if new_word == endWord:
                        return steps
                    if new_word in word_set:
                        word_set.discard(new_word)
                        queue.append(new_word)
    return 0


# ============================================================
# APPROACH 3: BEST - Bidirectional BFS
# Time: O(N * L * 26)  |  Space: O(N * L)
# Expand from both ends; dramatically reduces search space
# ============================================================
def best(beginWord: str, endWord: str, wordList: List[str]) -> int:
    word_set = set(wordList)
    if endWord not in word_set:
        return 0
    front = {beginWord}
    back = {endWord}
    word_set.discard(endWord)
    steps = 1
    while front:
        steps += 1
        # Always expand the smaller frontier
        if len(front) > len(back):
            front, back = back, front
        nxt = set()
        for word in front:
            for i in range(len(word)):
                for c in 'abcdefghijklmnopqrstuvwxyz':
                    new_word = word[:i] + c + word[i+1:]
                    if new_word in back:
                        return steps
                    if new_word in word_set:
                        word_set.discard(new_word)
                        nxt.add(new_word)
        front = nxt
    return 0


if __name__ == "__main__":
    tests = [
        ("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"], 5),
        ("hit", "cog", ["hot", "dot", "dog", "lot", "log"], 0),
        ("a", "c", ["a", "b", "c"], 2),
        ("hot", "dog", ["hot", "dog", "dot"], 3),
    ]
    for begin, end, wl, expected in tests:
        bf = brute_force(begin, end, wl)
        opt = optimal(begin, end, wl)
        be = best(begin, end, wl)
        status = "OK" if bf == opt == be == expected else "FAIL"
        print(f"[{status}] {begin}->{end}: Brute={bf}, Optimal={opt}, Best={be} (expected={expected})")
